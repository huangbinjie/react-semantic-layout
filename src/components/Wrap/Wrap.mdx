---
name: Wrap
menu: components
---

import { Playground, Props } from 'docz'
import Container from '../Container'
import Wrap from '../Wrap'
import Center from '../Center'

## Wrap

提供灵活的换行布局的组件。使用 flexbox + flex-wrap 实现自适应的换行布局。

Wrap 组件专注于提供简单灵活的换行布局，支持自定义间距和对齐方式。如果需要固定的网格布局，请使用 Grid 组件。

### 基本使用

默认支持自动换行：

<Playground>
  <Wrap mainAxisGap={10} crossAxisGap={5}>
    <Container width="100px" height="50px" color="#ececec">
      <Center>Item 1</Center>
    </Container>
    <Container width="100px" height="50px" color="coral">
      <Center>Item 2</Center>
    </Container>
    <Container width="100px" height="50px" color="blue">
      <Center style={{ color: '#fff' }}>Item 3</Center>
    </Container>
    <Container width="100px" height="50px" color="green">
      <Center style={{ color: '#fff' }}>Item 4</Center>
    </Container>  </Wrap>
</Playground>

### 自定义间距

支持设置不同的主轴和交叉轴间距：

<Playground>
  <Container width="300px">
    <Wrap mainAxisGap={20} crossAxisGap={15}>
      <Container width="100px" height="50px" color="#ececec">
        <Center>Item 1</Center>
      </Container>
      <Container width="100px" height="50px" color="coral">
        <Center>Item 2</Center>
      </Container>
      <Container width="100px" height="50px" color="blue">
        <Center style={{ color: '#fff' }}>Item 3</Center>
      </Container>
      <Container width="100px" height="50px" color="green">
        <Center style={{ color: '#fff' }}>Item 4</Center>
      </Container>
      <Container width="100px" height="50px" color="orange">
        <Center>Item 5</Center>
      </Container>
    </Wrap>
  </Container>
</Playground>

### 对齐方式

支持水平和垂直对齐：

<Playground>
  <Container height="200px" color="#f0f0f0">
    <Wrap 
      mainAxisAlignment="center" 
      crossAxisAlignment="center"
      crossAxisSize="max"
      mainAxisGap={10}
      crossAxisGap={10}
    >
      <Container width="80px" height="40px" color="coral">
        <Center>1</Center>
      </Container>
      <Container width="80px" height="40px" color="blue">
        <Center style={{ color: '#fff' }}>2</Center>
      </Container>
      <Container width="80px" height="40px" color="green">
        <Center style={{ color: '#fff' }}>3</Center>
      </Container>
    </Wrap>
  </Container>
</Playground>

## Props

<Props of={Wrap} />
