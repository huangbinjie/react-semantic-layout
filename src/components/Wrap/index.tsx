import React from 'react'
import { style as jss } from 'typestyle'
import classnames from 'classnames'
import { Alignment, decodeAlignment, decodeAxisSize } from '../../style'

export type WrapProps = {
  className?: string
  /** 水平对齐方式 */
  mainAxisAlignment?: Alignment
  /** 垂直对齐方式 */
  crossAxisAlignment?: Alignment
  /** 水平尺寸 */
  mainAxisSize?: 'min' | 'max'
  /** 垂直尺寸 */
  crossAxisSize?: 'min' | 'max'
  /** 水平间距 */
  mainAxisGap?: number | string
  /** 垂直间距 */
  crossAxisGap?: number | string
  style?: React.CSSProperties
}

/**
 * Wrap 组件，提供灵活的换行布局
 * 使用 flexbox + flex-wrap 实现自适应的换行布局
 */
export default function Wrap({
  className = '',
  mainAxisAlignment = 'start',
  crossAxisAlignment = 'start',
  mainAxisSize = 'max',
  crossAxisSize = 'min',
  mainAxisGap = 0,
  crossAxisGap = 0,
  children,
  ...restProps
}: React.PropsWithChildren<WrapProps>) {
  const mainAxisGapValue =
    typeof mainAxisGap === 'number' ? `${mainAxisGap}px` : mainAxisGap
  const crossAxisGapValue =
    typeof crossAxisGap === 'number' ? `${crossAxisGap}px` : crossAxisGap
  // 简单的 flex-wrap 布局
  const flexStyles: any = {
    display: 'flex',
    flexWrap: 'wrap',
    width: decodeAxisSize(mainAxisSize),
    height: decodeAxisSize(crossAxisSize),
    justifyContent: decodeAlignment(mainAxisAlignment),
    alignContent: decodeAlignment(crossAxisAlignment),
    gap: `${crossAxisGapValue} ${mainAxisGapValue}`,
  }

  const classname = jss(flexStyles)

  return (
    <div
      debug-label="Wrap"
      className={classnames(classname, className)}
      {...restProps}
    >
      {children}
    </div>
  )
}
