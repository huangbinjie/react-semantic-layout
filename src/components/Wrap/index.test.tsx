import React from 'react'
import renderer from 'react-test-renderer'

import Wrap from '.'

describe('Wrap', () => {
  test('basic', () => {
    const tree = renderer.create(<Wrap>empty content</Wrap>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with gaps only', () => {
    const tree = renderer
      .create(
        <Wrap mainAxisGap={15} crossAxisGap={10}>
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </Wrap>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with alignment', () => {
    const tree = renderer
      .create(
        <Wrap
          mainAxisAlignment="center"
          crossAxisAlignment="end"
          mainAxisGap={10}
          crossAxisGap={5}
        >
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
          <div>Item 4</div>
        </Wrap>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('with string gap values', () => {
    const tree = renderer
      .create(
        <Wrap mainAxisGap="1rem" crossAxisGap="0.5rem">
          <div>Item 1</div>
          <div>Item 2</div>
          <div>Item 3</div>
        </Wrap>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
