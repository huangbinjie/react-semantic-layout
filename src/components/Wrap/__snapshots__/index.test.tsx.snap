// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrap basic 1`] = `
<div
  className="f1pg3ywt"
  debug-label="Wrap"
>
  empty content
</div>
`;
